hibernate-core/src/main/java/org/hibernate/boot/registry/classloading/internal/ClassLoaderServiceImpl.java

```diff
@@ -157,19 +137,96 @@ private static ClassLoader locateTCCL() {
 		}
 	}
 
+	@Override
+	public TCCLLookupBehavior getTTCLLookupBehavior() {
+		return getAggregatedClassLoader().tcclLookupBehavior;
+	}
+
+	@Override
+	public void setTCCLLookupBehavior(TCCLLookupBehavior behavior) {
+		getAggregatedClassLoader().tcclLookupBehavior = behavior;
+	}
+
 	private static class AggregatedClassLoader extends ClassLoader {
 		private final ClassLoader[] individualClassLoaders;
+		private volatile TCCLLookupBehavior tcclLookupBehavior = TCCLLookupBehavior.AFTER;
 
 		private AggregatedClassLoader(final LinkedHashSet<ClassLoader> orderedClassLoaderSet) {
 			super( null );
 			individualClassLoaders = orderedClassLoaderSet.toArray( new ClassLoader[orderedClassLoaderSet.size()] );
 		}
 
+		private Iterator<ClassLoader> newClassLoaderIterator() {
+			final ClassLoader sysClassLoader = locateSystemClassLoader();	
+			final ClassLoader threadClassLoader = locateTCCL();
+			
+			final TCCLLookupBehavior behavior;
+			if ( threadClassLoader == null ) {
+				behavior = TCCLLookupBehavior.NEVER;
+			} 
+			else {
+				behavior = tcclLookupBehavior;
+			}
+			
+			return new Iterator<ClassLoader>() {
+				private boolean tcclReturned = false;
+				private boolean sysclReturned = false;
+				private int currentIndex = 0;
+
+				private ClassLoader nextClassLoader;
+				
+				@Override
+				public boolean hasNext() {
+					if ( nextClassLoader != null ) {
+						return true;
+					}
+					
+					if ( currentIndex == 0 && behavior == TCCLLookupBehavior.BEFORE && !tcclReturned ) {
+						tcclReturned = true;
+						nextClassLoader = threadClassLoader;
+						return true;
+					}
+					
+					if ( currentIndex < individualClassLoaders.length ) {
+						nextClassLoader = individualClassLoaders[ currentIndex ];
+						++currentIndex;
+						return true;
+					}
+					
+					if ( behavior == TCCLLookupBehavior.AFTER && !tcclReturned ) {
+						tcclReturned = true;
+						nextClassLoader = threadClassLoader;
+						return true;
+					}
+					
+					if ( !sysclReturned ) {
+						sysclReturned = true;
+						nextClassLoader = sysClassLoader;
+						return true;
+					}
+					
+					return false;
+				}
+
+				@Override
+				public ClassLoader next() {
```

Comment:
Why just 2 iterators?  Are you thinking one or the other would handle `NEVER` implicitly?